### PR TITLE
Provide better error messages for subst failures

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
Because of #4566 , we found out that we need better error messages.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

**Review by valentin-krasontovitsch — CHANGES_REQUESTED** [review]
can we make it better? give more context? like the whole line?

**Comment by valentin-krasontovitsch** [comment]
looking at the code, i see that in `config_parser.cpp`, we try to catch the define error, and append it to the parse errors list - but in testing a run, ert just crashes with the runtime error rather than reporting a parse error. is this because i'm trying to create the infinite loop inside arguments of the forward model keyword?

**Comment by valentin-krasontovitsch** [comment]
python cannot catch a runtime error thrown in C++ code, can it?

**Comment by oysteoh** [comment]
Seems like it works as intended. My only input would be to slightly modify of of the messages to make it more clear, but i would also accept the current message as good enough :)

File: src/clib/lib/res_util/subst_list.cpp [hunk 1]
```diff
@@ -359,8 +361,9 @@ char *subst_list_alloc_filtered_string(const subst_list_type *subst_list,
         }
 
         if (iterations >= max_iterations) {
-            throw std::runtime_error(
-                "Reached max iterations while trying to resolve defines");
+            throw std::runtime_error(fmt::format(
+                "Reached max iterations while trying to resolve defines in {}",
```
Comment (oysteoh) [inline src/clib/lib/res_util/subst_list.cpp:365]: Feel this sentence is not fully explaining what is going on - what about `what(): Reached max iterations while trying to resolve DEFINE for string <RMS_PROJECT>`

**Review by oysteoh — APPROVED** [review]

**Comment by oysteoh** [comment]
Very nice! :rocket:

**Comment by eivindjahren** [comment]
> python cannot catch a runtime error thrown in C++ code, can it?

Yes, this is the general problem. Attempting to communicate the error up the chain is going to be painful so I think CLI is as good as it gets.

**Comment by valentin-krasontovitsch** [comment]
turns out if we expose the c/c++ functions through pybind, and throw c++ exceptions, the exceptions get automatically propagated to python and can be caught and handled there. thanks to @pinkwah for the insights! i've created a separate PR where we can try to add some context to the error from further up the calling - #4581

**Review by valentin-krasontovitsch — APPROVED** [review]
looks good!

**Comment by oysteoh** [comment]
Jenkins: Test this please!

**Pull request merged**